Add logging to the assembler

Overcastan opened the pull request:
This PR implements [Tracing](https://docs.rs/tracing/latest/tracing/index.html#) logger for the Miden VM.

TODO: 
- [x] Update changelog

bobbinth commented on assembly/src/ast/mod.rs:
```diff
@@ -938,3 +941,17 @@ fn sort_procs_into_vec(proc_map: LocalProcMap) -> Vec<ProcedureAst> {
+
+    for lib in import_lib_paths {
+        if !invoked_procs_paths.contains(&lib) {
+            log::warn!("warning: unused import: \"{}\"", lib);
```
is it also possible to print out the path of the module in which the unused import is found?

bobbinth commented on miden/src/cli/data.rs:
```diff
@@ -406,6 +406,7 @@ impl ProgramFile {
         L: Library,
     {
         print!("Compiling program... ");
+        stdout().flush().expect("Couldn't flush stdout");
```
Why is `.flush()` needed here?

Also, we should probably replace all `print!` and `println!` statements with `info!` (this might require some updates to how we structure messages as I'm not sure we can print on the same line with `info!`).

Overcastan replied:
I was using `.flush()` here to be sure that the message of the logger will be exactly in the desired place. I want to print this info between "Compiling program... " and "done" strings, but since `print!` doesn't flush the output stream, the logger message appears to be after "Compiling program... done" line. 

But it shouldn't be a problem after I change all `print`s to the `info`s, so I'll be able to remove this `flush`.

bobbinth commented on assembly/src/assembler/mod.rs:
```diff
@@ -264,6 +264,13 @@ impl Assembler {
 
+        // log the completion of the module
+        #[cfg(feature = "std")]
+        log::debug!(
+            "\n - Module \"{}\" has been compiled",
+            path.unwrap_or(&LibraryPath::anon_path())
+        );
```
nit: It would be nice to add a timer, and then I'd probably change the message to "Compiled xyz module in 1.23 ms"

Also, I wonder if we should use `trace!` rather than `debug!` here.

bobbinth commented on miden/src/main.rs:
```diff
@@ -53,6 +54,16 @@ pub fn main() {
 
+    // configure logging
+    // if logging level is not specified, set level to "warn"
+    if std::env::var("MIDEN_LOG").is_err() {
+        std::env::set_var("MIDEN_LOG", "warn");
+    }
```
Now thinking about this, I'd probably set the default level to "info".

bobbinth reviewed (changes requested):
Looks good! Thank you! I left a few comments inline. Also, I think we should log parsing as well (not just compilation). E.g., "Parsed xyz module in 2.34 ms".

Lastly, could you upload a screenshot of how the output looks like (just one or two examples).

Overcastan commented:
I am not sure that logging of parsing completion will make sense for us, because at the parsing stage we don't have information about name or path of the module, we can obtain this info only at compiling stage. So I can log only something like "Parsed module in 2.34 ms", without specifying the name or the path of the module.

Overcastan commented:
Executing this program 
```
use.module::path::two
use.bar::baz

begin 
    exec.two::foo
end
```
we will get this output without logging level specification:
```
% target/optimized/miden run -a miden/examples/logger/logger.masm -i miden/examples/logger/logger.inputs -l miden/examples/logger/logger.masl
============================================================
Run program
============================================================
Reading library file `miden/examples/logger/logger.masl`
Reading program file `miden/examples/logger/logger.masm`
Parsing program... 
warning: unused imports: "bar::baz"
done (0 ms)
Compiling program... 
done (4 ms)
Reading input file `miden/examples/logger/logger.inputs`
Executing program with hash 8bcf4aab5eaba5f76fed175478a14f84cc61f1c47fcec5fdb138fa38bfff7595... done (0 ms)
Output: [2, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
VM cycles: 39 extended to 64 steps (39% padding).
├── Stack rows: 6
├── Range checker rows: 39
└── Chiplets rows: 9
    ├── Hash chiplet rows: 8
    ├── Bitwise chiplet rows: 0
    ├── Memory chiplet rows: 0
    └── Kernel ROM rows: 0
```
and this output by specifying level to be `trace`:
```
% MIDEN_LOG=trace target/optimized/miden run -a miden/examples/logger/logger.masm -i miden/examples/logger/logger.inputs -l miden/examples/logger/logger.masl
============================================================
Run program
============================================================
Reading library file `miden/examples/logger/logger.masl`
Reading program file `miden/examples/logger/logger.masm`
Parsing program... 
warning: unused imports: "bar::baz"
done (0 ms)
Compiling program... 
- Compiled "module::path::one" module in 0 ms
- Compiled "module::path::two" module in 0 ms
done (2 ms)
Reading input file `miden/examples/logger/logger.inputs`
Executing program with hash 8bcf4aab5eaba5f76fed175478a14f84cc61f1c47fcec5fdb138fa38bfff7595... done (0 ms)
Output: [2, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
VM cycles: 39 extended to 64 steps (39% padding).
├── Stack rows: 6
├── Range checker rows: 39
└── Chiplets rows: 9
    ├── Hash chiplet rows: 8
    ├── Bitwise chiplet rows: 0
    ├── Memory chiplet rows: 0
    └── Kernel ROM rows: 0
```

Used `logger` library is just a dummy with two modules, where only module `one` contains function `foo`, which I use in the example.

Overcastan commented:
One thing I don't like in my implementation is that I have to put the `done (* ms)` string to the new line after parsing and compiling regardless of whether we print warnings or information about compiled modules between them.

bobbinth reviewed (changes requested):
Thank you! Looks good! I left some comments inline.

> One thing I don't like in my implementation is that I have to put the `done (* ms)` string to the new line after parsing and compiling regardless of whether we print warnings or information about compiled modules between them.

Yes, agreed. I think we should replace `done(* ms)` with more meaningful text.

Another option is to use a more sophisticated logging library - but I'm not sure it is worth it.

bobbinth commented on miden/src/cli/prove.rs:
```diff
@@ -61,15 +61,9 @@ impl ProveCmd {
-            .format(|buf, record| writeln!(buf, "{}", record.args()))
-            .filter_level(log::LevelFilter::Debug)
-            .init();
+        log::info!("============================================================\n");
+        log::info!("Prove program\n");
+        log::info!("============================================================\n");
```
After thinking a bit more, I'm wondering if things like this should remain `println!`. One can argue that this is not really logging output but an expected output of the command. That is, we should print out this and other statements in this file regardless of whether logging is enabled or not.

Overcastan commented:
Here is updated output for proving with default logging level:
```
% target/optimized/miden prove -a miden/examples/logger/logger.masm -i miden/examples/logger/logger.inputs -l miden/examples/logger/lib/logger_lib.masl                
============================================================
Prove program
============================================================
Reading library file `miden/examples/logger/lib/logger_lib.masl`
Reading program file `miden/examples/logger/logger.masm`
Parsing program...
warning: unused imports: "logger_lib::module2"
Parsed the program in 0 ms
Compiling program...
Compiled the program in 3 ms
Reading input file `miden/examples/logger/logger.inputs`
Proving program with hash 74184a73e531e907b5568906c1c2b53732bebcdcd7e197892694aba1d6ce197d...
Program with hash 74184a73e531e907b5568906c1c2b53732bebcdcd7e197892694aba1d6ce197d proved in 15 ms
Creating proof file `miden/examples/logger/logger.proof`
Writing data to proof file - size 31 KB
Creating output file `miden/examples/logger/logger.outputs`
Writing data to output file
Output: [3, 7, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
```

Here is the output with `trace` logging level: 
```
% MIDEN_LOG=trace target/optimized/miden prove -a miden/examples/logger/logger.masm -i miden/examples/logger/logger.inputs -l miden/examples/logger/lib/logger_lib.masl
============================================================
Prove program
============================================================
Reading library file `miden/examples/logger/lib/logger_lib.masl`
Reading program file `miden/examples/logger/logger.masm`
Parsing program...
warning: unused imports: "logger_lib::module2"
Parsed the program in 0 ms
Compiling program...
- Compiled "logger_lib::module1" module in 0 ms
Compiled the program in 2 ms
Reading input file `miden/examples/logger/logger.inputs`
Proving program with hash 74184a73e531e907b5568906c1c2b53732bebcdcd7e197892694aba1d6ce197d...
Generated execution trace of 70 columns and 64 steps (26% padded) in 0 ms
Built domain of 2^9 elements in 0 ms
Extended execution trace of 70 columns from 2^6 to 2^9 steps (8x blowup) in 0 ms
Computed execution trace commitment (Merkle tree of depth 9) in 0 ms
Built auxiliary trace segment of 7 columns and 2^6 steps in 0 ms
Extended execution trace of 7 columns from 2^6 to 2^9 steps (8x blowup) in 0 ms
Computed execution trace commitment (Merkle tree of depth 9) in 0 ms
Evaluated constraints over domain of 2^9 elements in 1 ms
Converted constraint evaluations into 8 composition polynomial columns of degree 63 in 0 ms
Evaluated 8 composition polynomial columns over LDE domain (2^9 elements) in 0 ms
Computed constraint evaluation commitment (Merkle tree of depth 9) in 0 ms
Built DEEP composition polynomial of degree 62 in 6 ms
Evaluated DEEP composition polynomial over LDE domain (2^9 elements) in 0 ms
Computed 0 FRI layers from composition polynomial evaluations in 0 ms
Determined 27 unique query positions in 1 ms
Built proof object in 0 ms
Program with hash 74184a73e531e907b5568906c1c2b53732bebcdcd7e197892694aba1d6ce197d proved in 13 ms
Creating proof file `miden/examples/logger/logger.proof`
Writing data to proof file - size 31 KB
Creating output file `miden/examples/logger/logger.outputs`
Writing data to output file
Output: [3, 7, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
```

Here is the output with logger turned off:
```
% MIDEN_LOG=off target/optimized/miden prove -a miden/examples/logger/logger.masm -i miden/examples/logger/logger.inputs -l miden/examples/logger/lib/logger_lib.masl
============================================================
Prove program
============================================================
Proving program with hash 74184a73e531e907b5568906c1c2b53732bebcdcd7e197892694aba1d6ce197d...
Program with hash 74184a73e531e907b5568906c1c2b53732bebcdcd7e197892694aba1d6ce197d proved in 19 ms
Output: [3, 7, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
```

After I added the unused module warning, library parsing and compilation started to look like so:
```
% target/optimized/miden bundle miden/examples/logger/lib -n logger_lib
============================================================
Build library
============================================================
warning: unused imports: "logger_lib::module1"
Built library logger_lib
```

For some reason capital letter `W` is printed with weird space after it. This may be related to the font used in the terminal, but I didn't get the exact reason yet.

<img width="124" alt="Screenshot 2023-12-14 at 6 19 26 PM" src="https://github.com/0xPolygonMiden/miden-vm/assets/43154072/d59926b2-ac92-4914-80cd-50bdf7b0109a">

Overcastan commented:
Currently this is how log looks like:
```
% MIDEN_LOG=trace target/optimized/miden run -a miden/examples/logger/logger.masm -i miden/examples/logger/logger.inputs -l miden/examples/logger/lib/logger_lib.masl
INFO      ===== Run program ===== [ 4.41ms | 3.85% / 100.00% ]
INFO     ┝━ Reading library files [ 102µs | 0.27% / 2.33% ]
INFO     │  ┕━ Reading library file [ 90.4µs | 2.05% ] path: miden/examples/logger/lib/logger_lib.masl
INFO     ┝━ Reading program file [ 81.3µs | 0.88% / 1.85% ] path: miden/examples/logger/logger.masm
INFO     │  ┕━ Parsing program [ 42.7µs | 0.97% ]
WARN     │     ┕━ 🚧 [warn]: warning: unused import: "logger_lib::module2"
INFO     ┝━ Compiling program [ 3.35ms | 74.15% / 76.04% ]
INFO     │  ┕━ Compile AST [ 83.4µs | 1.14% / 1.89% ]
TRACE    │     ┕━ Compiling module [ 33.0µs | 0.75% ] module: "logger_lib::module1"
INFO     ┝━ Reading input file [ 67.0µs | 1.52% ]
INFO     ┝━ Executing program [ 635µs | 14.41% ] hash: 74184a73e531e907b5568906c1c2b53732bebcdcd7e197892694aba1d6ce197d
INFO     ┝━ ｉ [info]: Output: [3, 7, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
INFO     ┕━ ｉ [info]: VM cycles: 47 extended to 64 steps (26% padding).
                       ├── Stack rows: 11
                       ├── Range checker rows: 47
                       └── Chiplets rows: 9
                           ├── Hash chiplet rows: 8
                           ├── Bitwise chiplet rows: 0
                           ├── Memory chiplet rows: 0
                           └── Kernel ROM rows: 0
```

```
% target/optimized/miden prove -a miden/examples/logger/logger.masm -i miden/examples/logger/logger.inputs -l miden/examples/logger/lib/logger_lib.masl         
INFO      ===== Prove program ===== [ 19.3ms | 0.79% / 100.00% ]
INFO     ┝━ Reading library files [ 84.5µs | 0.06% / 0.44% ]
INFO     │  ┕━ Reading library file [ 72.7µs | 0.38% ] path: miden/examples/logger/lib/logger_lib.masl
INFO     ┝━ Reading program file [ 69.2µs | 0.14% / 0.36% ] path: miden/examples/logger/logger.masm
INFO     │  ┕━ Parsing program [ 41.5µs | 0.21% ]
WARN     │     ┕━ 🚧 [warn]: warning: unused import: "logger_lib::module2"
INFO     ┝━ Compiling program [ 3.41ms | 17.30% / 17.67% ]
INFO     │  ┕━ Compile AST [ 70.3µs | 0.36% ]
INFO     ┝━ Reading input file [ 66.4µs | 0.34% ]
INFO     ┝━ Proving program [ 14.9ms | 74.43% / 76.93% ]
INFO     │  ┕━ Executing program [ 483µs | 2.50% ] hash: 74184a73e531e907b5568906c1c2b53732bebcdcd7e197892694aba1d6ce197d
INFO     ┝━ Writing data to proof file [ 240µs | 1.24% ] path: miden/examples/logger/logger.proof | size: "31 KB"
INFO     ┝━ Writing data to output file [ 430µs | 2.23% ] path: miden/examples/logger/logger.outputs
INFO     ┕━ ｉ [info]: Output: [3, 7, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
```

```
% MIDEN_LOG=trace target/optimized/miden prove -a miden/examples/logger/logger.masm -i miden/examples/logger/logger.inputs -l miden/examples/logger/lib/logger_lib.masl 
INFO      ===== Prove program ===== [ 20.8ms | 0.73% / 100.00% ]
INFO     ┝━ Reading library files [ 128µs | 0.06% / 0.62% ]
INFO     │  ┕━ Reading library file [ 116µs | 0.56% ] path: miden/examples/logger/lib/logger_lib.masl
INFO     ┝━ Reading program file [ 85.7µs | 0.21% / 0.41% ] path: miden/examples/logger/logger.masm
INFO     │  ┕━ Parsing program [ 42.8µs | 0.21% ]
WARN     │     ┕━ 🚧 [warn]: warning: unused import: "logger_lib::module2"
INFO     ┝━ Compiling program [ 3.37ms | 15.85% / 16.25% ]
INFO     │  ┕━ Compile AST [ 83.2µs | 0.24% / 0.40% ]
TRACE    │     ┕━ Compiling module [ 33.1µs | 0.16% ] module: "logger_lib::module1"
INFO     ┝━ Reading input file [ 62.8µs | 0.30% ]
INFO     ┝━ Proving program [ 15.3ms | 71.41% / 73.60% ]
INFO     │  ┝━ Executing program [ 454µs | 2.19% ] hash: 74184a73e531e907b5568906c1c2b53732bebcdcd7e197892694aba1d6ce197d
DEBUG    │  ┝━ 🐛 [debug]: Generated execution trace of 70 columns and 64 steps (26% padded) in 0 ms | log.target: "miden_prover" | log.module_path: "miden_prover" | log.file: "prover/src/lib.rs" | log.line: 71
DEBUG    │  ┝━ 🐛 [debug]: Built domain of 2^9 elements in 0 ms | log.target: "winter_prover" | log.module_path: "winter_prover" | log.file: "/Users/fumuran/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winter-prover-0.7.1/src/lib.rs" | log.line: 264
DEBUG    │  ┝━ 🐛 [debug]: Extended execution trace of 70 columns from 2^6 to 2^9 steps (8x blowup) in 0 ms | log.target: "winter_prover::trace::trace_lde::default" | log.module_path: "winter_prover::trace::trace_lde::default" | log.file: "/Users/fumuran/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winter-prover-0.7.1/src/trace/trace_lde/default/mod.rs" | log.line: 242
DEBUG    │  ┝━ 🐛 [debug]: Computed execution trace commitment (Merkle tree of depth 9) in 0 ms | log.target: "winter_prover::trace::trace_lde::default" | log.module_path: "winter_prover::trace::trace_lde::default" | log.file: "/Users/fumuran/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winter-prover-0.7.1/src/trace/trace_lde/default/mod.rs" | log.line: 256
DEBUG    │  ┝━ 🐛 [debug]: Built auxiliary trace segment of 7 columns and 2^6 steps in 0 ms | log.target: "winter_prover" | log.module_path: "winter_prover" | log.file: "/Users/fumuran/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winter-prover-0.7.1/src/lib.rs" | log.line: 297
DEBUG    │  ┝━ 🐛 [debug]: Extended execution trace of 7 columns from 2^6 to 2^9 steps (8x blowup) in 0 ms | log.target: "winter_prover::trace::trace_lde::default" | log.module_path: "winter_prover::trace::trace_lde::default" | log.file: "/Users/fumuran/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winter-prover-0.7.1/src/trace/trace_lde/default/mod.rs" | log.line: 242
DEBUG    │  ┝━ 🐛 [debug]: Computed execution trace commitment (Merkle tree of depth 9) in 0 ms | log.target: "winter_prover::trace::trace_lde::default" | log.module_path: "winter_prover::trace::trace_lde::default" | log.file: "/Users/fumuran/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winter-prover-0.7.1/src/trace/trace_lde/default/mod.rs" | log.line: 256
DEBUG    │  ┝━ 🐛 [debug]: Evaluated constraints over domain of 2^9 elements in 2 ms | log.target: "winter_prover" | log.module_path: "winter_prover" | log.file: "/Users/fumuran/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winter-prover-0.7.1/src/lib.rs" | log.line: 337
DEBUG    │  ┝━ 🐛 [debug]: Converted constraint evaluations into 8 composition polynomial columns of degree 63 in 0 ms | log.target: "winter_prover" | log.module_path: "winter_prover" | log.file: "/Users/fumuran/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winter-prover-0.7.1/src/lib.rs" | log.line: 504
DEBUG    │  ┝━ 🐛 [debug]: Evaluated 8 composition polynomial columns over LDE domain (2^9 elements) in 0 ms | log.target: "winter_prover" | log.module_path: "winter_prover" | log.file: "/Users/fumuran/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winter-prover-0.7.1/src/lib.rs" | log.line: 519
DEBUG    │  ┝━ 🐛 [debug]: Computed constraint evaluation commitment (Merkle tree of depth 9) in 0 ms | log.target: "winter_prover" | log.module_path: "winter_prover" | log.file: "/Users/fumuran/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winter-prover-0.7.1/src/lib.rs" | log.line: 532
DEBUG    │  ┝━ 🐛 [debug]: Built DEEP composition polynomial of degree 62 in 7 ms | log.target: "winter_prover" | log.module_path: "winter_prover" | log.file: "/Users/fumuran/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winter-prover-0.7.1/src/lib.rs" | log.line: 392
DEBUG    │  ┝━ 🐛 [debug]: Evaluated DEEP composition polynomial over LDE domain (2^9 elements) in 0 ms | log.target: "winter_prover" | log.module_path: "winter_prover" | log.file: "/Users/fumuran/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winter-prover-0.7.1/src/lib.rs" | log.line: 413
DEBUG    │  ┝━ 🐛 [debug]: Computed 0 FRI layers from composition polynomial evaluations in 0 ms | log.target: "winter_prover" | log.module_path: "winter_prover" | log.file: "/Users/fumuran/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winter-prover-0.7.1/src/lib.rs" | log.line: 425
DEBUG    │  ┝━ 🐛 [debug]: Determined 26 unique query positions in 1 ms | log.target: "winter_prover" | log.module_path: "winter_prover" | log.file: "/Users/fumuran/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winter-prover-0.7.1/src/lib.rs" | log.line: 441
DEBUG    │  ┕━ 🐛 [debug]: Built proof object in 0 ms | log.target: "winter_prover" | log.module_path: "winter_prover" | log.file: "/Users/fumuran/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winter-prover-0.7.1/src/lib.rs" | log.line: 471
INFO     ┝━ Writing data to proof file [ 1.21ms | 5.85% ] path: miden/examples/logger/logger.proof | size: "31 KB"
INFO     ┝━ Writing data to output file [ 464µs | 2.24% ] path: miden/examples/logger/logger.outputs
INFO     ┕━ ｉ [info]: Output: [3, 7, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
```

There are **a lot** of additional information while calling `prove` with `debug` or `trace` logging level, but I decided to leave it as it is, since this information could be helpful for debugging (I already used it once).

bobbinth reviewed (commented):
Looks cool! Thank you! I left a couple of minor comments inline - but there are two more general comments:

First, I think we should separate logging output from "regular" output. Basically, regular output is something we'd like to see even if we turn the logging off completely. For example, for `run` command, it could be something like:
```
===============================================================================
Run program: miden/examples/fib/fib.masm
-------------------------------------------------------------------------------
Executed program with hash 78d31702eb946e1817e3c0881fcb3739562f08fbddf202de2eae241b9968ab3b in 1 ms
Output: [11112721240812633725]
VM cycles: 3045 extended to 4096 steps (25% padding).
├── Stack rows: 3045
├── Range checker rows: 39
└── Chiplets rows: 337
    ├── Hash chiplet rows: 336
    ├── Bitwise chiplet rows: 0
    ├── Memory chiplet rows: 0
    └── Kernel ROM rows: 0
```
So, I would print the above using regular `println` and would use logging for everything else. If we do this, we can set logging default back to `WARN`. Basically, if people want to see more detail, they can enable lower levels of logging. With logging enabled, it might looks something like this:
```
===============================================================================
Run program: miden/examples/fib/fib.masm
-------------------------------------------------------------------------------
INFO     ┝━ Reading library files [ 102µs | 0.27% / 2.33% ]
INFO     │  ┕━ Reading library file [ 90.4µs | 2.05% ] path: miden/examples/logger/lib/logger_lib.masl
INFO     ┝━ Reading program file [ 81.3µs | 0.88% / 1.85% ] path: miden/examples/logger/logger.masm
INFO     │  ┕━ Parsing program [ 42.7µs | 0.97% ]
WARN     │     ┕━ 🚧 [warn]: warning: unused import: "logger_lib::module2"
INFO     ┝━ Compiling program [ 3.35ms | 74.15% / 76.04% ]
INFO     │  ┕━ Compile AST [ 83.4µs | 1.14% / 1.89% ]
TRACE    │     ┕━ Compiling module [ 33.0µs | 0.75% ] module: "logger_lib::module1"
INFO     ┝━ Reading input file [ 67.0µs | 1.52% ]
INFO     ┕━ Executing program [ 635µs | 14.41% ]
Executed program with hash 78d31702eb946e1817e3c0881fcb3739562f08fbddf202de2eae241b9968ab3b in 1 ms
Output: [11112721240812633725]
VM cycles: 3045 extended to 4096 steps (25% padding).
├── Stack rows: 3045
├── Range checker rows: 39
└── Chiplets rows: 337
    ├── Hash chiplet rows: 336
    ├── Bitwise chiplet rows: 0
    ├── Memory chiplet rows: 0
    └── Kernel ROM rows: 0
```

We should do the same for other commands as well (obviously, what we consider "regular" will differ from command to command).

Second, for log output formatting, could we do the following:
- Not show the left column with `INFO`, `WARN` etc. labels?
- For winterfell output, note show the very verbose file location etc. to the right of the message?

bobbinth commented:
So, for the `run` command, it would be awesome if the output (with logging enabled) looked something like:
```
===============================================================================
Run program: miden/examples/fib/fib.masm
-------------------------------------------------------------------------------
┝━ Reading library files [ 102µs | 0.27% / 2.33% ]
│  ┕━ Reading library file [ 90.4µs | 2.05% ] path: miden/examples/logger/lib/logger_lib.masl
┝━ Reading program file [ 81.3µs | 0.88% / 1.85% ] path: miden/examples/logger/logger.masm
│  ┕━ Parsing program [ 42.7µs | 0.97% ]
│     ┕━ 🚧 [warn]: warning: unused import: "logger_lib::module2"
┝━ Compiling program [ 3.35ms | 74.15% / 76.04% ]
│  ┕━ Compile AST [ 83.4µs | 1.14% / 1.89% ]
│     ┕━ Compiling module [ 33.0µs | 0.75% ] module: "logger_lib::module1"
┝━ Reading input file [ 67.0µs | 1.52% ]
┕━ Executing program [ 635µs | 14.41% ]
Executed program with hash 78d31702eb946e1817e3c0881fcb3739562f08fbddf202de2eae241b9968ab3b in 1 ms
Output: [11112721240812633725]
VM cycles: 3045 extended to 4096 steps (25% padding).
├── Stack rows: 3045
├── Range checker rows: 39
└── Chiplets rows: 337
    ├── Hash chiplet rows: 336
    ├── Bitwise chiplet rows: 0
    ├── Memory chiplet rows: 0
    └── Kernel ROM rows: 0
```

Overcastan commented:
After the update output looks like this: 

- `run` with disabled logging: 
```
% MIDEN_LOG=off target/optimized/miden run -i miden/examples/logger/logger.inputs -a miden/examples/logger/logger.masm -l miden/examples/logger/lib/logger_lib.masl 
===============================================================================
Run program: miden/examples/logger/logger.masm
-------------------------------------------------------------------------------
Executed the program with hash 74184a73e531e907b5568906c1c2b53732bebcdcd7e197892694aba1d6ce197d in 6 ms
Output: [3, 7, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
VM cycles: 47 extended to 64 steps (26% padding).
├── Stack rows: 11
├── Range checker rows: 47
└── Chiplets rows: 9
    ├── Hash chiplet rows: 8
    ├── Bitwise chiplet rows: 0
    ├── Memory chiplet rows: 0
    └── Kernel ROM rows: 0
```

- `run` with default (`warn`) logging level:
```
% target/optimized/miden run -i miden/examples/logger/logger.inputs -a miden/examples/logger/logger.masm -l miden/examples/logger/lib/logger_lib.masl 
===============================================================================
Run program: miden/examples/logger/logger.masm
-------------------------------------------------------------------------------
WARN     🚧 [warn]: unused import: "logger_lib::module2"
Executed the program with hash 74184a73e531e907b5568906c1c2b53732bebcdcd7e197892694aba1d6ce197d in 2 ms
Output: [3, 7, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
VM cycles: 47 extended to 64 steps (26% padding).
├── Stack rows: 11
├── Range checker rows: 47
└── Chiplets rows: 9
    ├── Hash chiplet rows: 8
    ├── Bitwise chiplet rows: 0
    ├── Memory chiplet rows: 0
    └── Kernel ROM rows: 0
```

- `run` with `trace` logging level:
```
% MIDEN_LOG=trace target/optimized/miden run -i miden/examples/logger/logger.inputs -a miden/examples/logger/logger.masm -l miden/examples/logger/lib/logger_lib.masl
===============================================================================
Run program: miden/examples/logger/logger.masm
-------------------------------------------------------------------------------
INFO     Running program [ 4.41ms | 2.89% / 100.00% ]
INFO     ┝━ Reading library files [ 114µs | 0.53% / 2.59% ]
INFO     │  ┕━ Reading library file [ 90.8µs | 2.06% ] path: miden/examples/logger/lib/logger_lib.masl
INFO     ┝━ Reading program file [ 98.9µs | 0.70% / 2.24% ] path: miden/examples/logger/logger.masm
INFO     │  ┕━ Parsing program [ 68.2µs | 1.55% ]
WARN     │     ┕━ 🚧 [warn]: unused import: "logger_lib::module2"
INFO     ┝━ Compiling program [ 3.41ms | 74.10% / 77.34% ]
INFO     │  ┕━ Compile AST [ 143µs | 2.28% / 3.25% ]
TRACE    │     ┕━ Compiling module [ 42.7µs | 0.97% ] module: "logger_lib::module1"
INFO     ┝━ Reading input file [ 57.4µs | 1.30% ]
INFO     ┕━ Executing program [ 602µs | 13.64% ]
Executed the program with hash 74184a73e531e907b5568906c1c2b53732bebcdcd7e197892694aba1d6ce197d in 4 ms
Output: [3, 7, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
VM cycles: 47 extended to 64 steps (26% padding).
├── Stack rows: 11
├── Range checker rows: 47
└── Chiplets rows: 9
    ├── Hash chiplet rows: 8
    ├── Bitwise chiplet rows: 0
    ├── Memory chiplet rows: 0
    └── Kernel ROM rows: 0
```

Output of the `prove` and `verify` looks similar.

However, there are still several problems:
1. There is no way we can disable printing the logging level at the beginning of the line (`INFO`, `WARN` etc.) without rewriting the whole tree-like output: these levels are placed in the core of this tree. Potentially I can rewrite the whole tree, but I'm not sure that it is worth it. 
2. I shortened the unnecessarily detailed lines in the `DEBUG` output of the prover (the ones that were in the VM repo), but most of them are printed from the winterfell implementation of the prover. So to make them all short I need to implement Tracing in the winterfell too, but I'm not sure we want to do that now. Anyway, I can create an issue for that.

bobbinth commented on prover/src/lib.rs:
```diff
@@ -67,7 +68,8 @@ where
         / trace.trace_len_summary().padded_trace_len();
     #[cfg(feature = "std")]
-    debug!(
+    event!(
+        Level::DEBUG,
```
Let's change the level to `INFO` here.

bobbinth commented on prover/src/gpu.rs:
```diff
@@ -115,7 +115,8 @@ where
             get_evaluation_offsets::<E>(composition_poly.column_len(), blowup, domain.offset());
         let segments = build_segments(composition_poly.data(), domain.trace_twiddles(), &offsets);
-        debug!(
+        event!(
+            Level::DEBUG,
```
Let's change level to `INFO` here and in other places in this file.

bobbinth commented on miden/src/main.rs:
```diff
@@ -53,6 +55,17 @@ pub fn main() {
     // read command-line args
     let cli = Cli::parse();
 
+    // configure logging
+    // if logging level is not specified, set level to "info"
+    if std::env::var("MIDEN_LOG").is_err() {
+        std::env::set_var("MIDEN_LOG", "warn");
+    }
```
The comment here should be updated.

bobbinth commented on miden/src/cli/verify.rs:
```diff
@@ -22,9 +22,9 @@ pub struct VerifyCmd {
-        println!("Verify program");
-        println!("============================================================");
+        println!("===============================================================================");
+        println!("Verify program: {}", self.proof_file.display());
```
Let's change this to `Verifying proof: {}`.

bobbinth commented on miden/src/cli/verify.rs:
```diff
@@ -41,7 +41,7 @@ impl VerifyCmd {
         let proof = ProofFile::read(&Some(self.proof_file.clone()), &self.proof_file)?;
 
-        println!("verifying program...");
+        println!("Verifying program...");
```
I'm wondering if we need to print out this line. How does the output for this command look like?

bobbinth replied:
I think we should remove it.

bobbinth commented on miden/src/cli/prove.rs:
```diff
@@ -1,8 +1,9 @@
+
+use std::{path::PathBuf, time::Instant};
 
 // TODO check if clap is supporting automatic generation of list values of hash function
```
Should we remove this comment? Doesn't seem relevant any more.

bobbinth commented on miden/src/cli/prove.rs:
How does output for `prove` command look like?

Overcastan replied:
`prove` with logging turned off now looks like this: 
```
% MIDEN_LOG=off target/optimized/miden prove -a miden/examples/logger/logger.masm -i miden/examples/logger/logger.inputs -l miden/examples/logger/lib/logger_lib.masl
===============================================================================
Prove program: miden/examples/logger/logger.masm
-------------------------------------------------------------------------------
Proving program with hash 74184a73e531e907b5568906c1c2b53732bebcdcd7e197892694aba1d6ce197d...
Program with hash 74184a73e531e907b5568906c1c2b53732bebcdcd7e197892694aba1d6ce197d proved in 20 ms
Output: [3, 7, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
```
with default logging level looks like this: 
```
% target/optimized/miden prove -a miden/examples/logger/logger.masm -i miden/examples/logger/logger.inputs -l miden/examples/logger/lib/logger_lib.masl 
===============================================================================
Prove program: miden/examples/logger/logger.masm
-------------------------------------------------------------------------------
WARN     🚧 [warn]: unused import: "logger_lib::module2"
Proving program with hash 74184a73e531e907b5568906c1c2b53732bebcdcd7e197892694aba1d6ce197d...
Program with hash 74184a73e531e907b5568906c1c2b53732bebcdcd7e197892694aba1d6ce197d proved in 14 ms
Output: [3, 7, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
```
and, the most untidy one, `prove` with `trace` logging level looks like this: 
```
% MIDEN_LOG=trace target/optimized/miden prove -a miden/examples/logger/logger.masm -i miden/examples/logger/logger.inputs -l miden/examples/logger/lib/logger_lib.masl
===============================================================================
Prove program: miden/examples/logger/logger.masm
-------------------------------------------------------------------------------
INFO     load_data [ 3.41ms | 0.24% / 100.00% ]
INFO     ┝━ Reading library files [ 105µs | 0.28% / 3.08% ]
INFO     │  ┕━ Reading library file [ 95.6µs | 2.80% ] path: miden/examples/logger/lib/logger_lib.masl
INFO     ┝━ Reading program file [ 75.8µs | 0.88% / 2.22% ] path: miden/examples/logger/logger.masm
INFO     │  ┕━ Parsing program [ 45.6µs | 1.34% ]
WARN     │     ┕━ 🚧 [warn]: unused import: "logger_lib::module2"
INFO     ┝━ Compiling program [ 3.15ms | 90.04% / 92.43% ]
INFO     │  ┕━ Compile AST [ 81.4µs | 1.35% / 2.39% ]
TRACE    │     ┕━ Compiling module [ 35.3µs | 1.04% ] module: "logger_lib::module1"
INFO     ┕━ Reading input file [ 69.2µs | 2.03% ]
Proving program with hash 74184a73e531e907b5568906c1c2b53732bebcdcd7e197892694aba1d6ce197d...
INFO     Proving program [ 15.6ms | 96.91% / 100.00% ]
INFO     ┝━ Executing program [ 481µs | 3.09% ]
DEBUG    ┝━ 🐛 [debug]: Generated execution trace of 70 columns and 64 steps (26% padded) in 0 ms
DEBUG    ┝━ 🐛 [debug]: Built domain of 2^9 elements in 0 ms | log.target: "winter_prover" | log.module_path: "winter_prover" | log.file: "/Users/fumuran/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winter-prover-0.7.1/src/lib.rs" | log.line: 264
DEBUG    ┝━ 🐛 [debug]: Extended execution trace of 70 columns from 2^6 to 2^9 steps (8x blowup) in 1 ms | log.target: "winter_prover::trace::trace_lde::default" | log.module_path: "winter_prover::trace::trace_lde::default" | log.file: "/Users/fumuran/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winter-prover-0.7.1/src/trace/trace_lde/default/mod.rs" | log.line: 242
DEBUG    ┝━ 🐛 [debug]: Computed execution trace commitment (Merkle tree of depth 9) in 0 ms | log.target: "winter_prover::trace::trace_lde::default" | log.module_path: "winter_prover::trace::trace_lde::default" | log.file: "/Users/fumuran/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winter-prover-0.7.1/src/trace/trace_lde/default/mod.rs" | log.line: 256
DEBUG    ┝━ 🐛 [debug]: Built auxiliary trace segment of 7 columns and 2^6 steps in 0 ms | log.target: "winter_prover" | log.module_path: "winter_prover" | log.file: "/Users/fumuran/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winter-prover-0.7.1/src/lib.rs" | log.line: 297
DEBUG    ┝━ 🐛 [debug]: Extended execution trace of 7 columns from 2^6 to 2^9 steps (8x blowup) in 0 ms | log.target: "winter_prover::trace::trace_lde::default" | log.module_path: "winter_prover::trace::trace_lde::default" | log.file: "/Users/fumuran/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winter-prover-0.7.1/src/trace/trace_lde/default/mod.rs" | log.line: 242
DEBUG    ┝━ 🐛 [debug]: Computed execution trace commitment (Merkle tree of depth 9) in 0 ms | log.target: "winter_prover::trace::trace_lde::default" | log.module_path: "winter_prover::trace::trace_lde::default" | log.file: "/Users/fumuran/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winter-prover-0.7.1/src/trace/trace_lde/default/mod.rs" | log.line: 256
DEBUG    ┝━ 🐛 [debug]: Evaluated constraints over domain of 2^9 elements in 2 ms | log.target: "winter_prover" | log.module_path: "winter_prover" | log.file: "/Users/fumuran/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winter-prover-0.7.1/src/lib.rs" | log.line: 337
DEBUG    ┝━ 🐛 [debug]: Converted constraint evaluations into 8 composition polynomial columns of degree 63 in 0 ms | log.target: "winter_prover" | log.module_path: "winter_prover" | log.file: "/Users/fumuran/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winter-prover-0.7.1/src/lib.rs" | log.line: 504
DEBUG    ┝━ 🐛 [debug]: Evaluated 8 composition polynomial columns over LDE domain (2^9 elements) in 0 ms | log.target: "winter_prover" | log.module_path: "winter_prover" | log.file: "/Users/fumuran/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winter-prover-0.7.1/src/lib.rs" | log.line: 519
DEBUG    ┝━ 🐛 [debug]: Computed constraint evaluation commitment (Merkle tree of depth 9) in 0 ms | log.target: "winter_prover" | log.module_path: "winter_prover" | log.file: "/Users/fumuran/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winter-prover-0.7.1/src/lib.rs" | log.line: 532
DEBUG    ┝━ 🐛 [debug]: Built DEEP composition polynomial of degree 62 in 8 ms | log.target: "winter_prover" | log.module_path: "winter_prover" | log.file: "/Users/fumuran/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winter-prover-0.7.1/src/lib.rs" | log.line: 392
DEBUG    ┝━ 🐛 [debug]: Evaluated DEEP composition polynomial over LDE domain (2^9 elements) in 0 ms | log.target: "winter_prover" | log.module_path: "winter_prover" | log.file: "/Users/fumuran/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winter-prover-0.7.1/src/lib.rs" | log.line: 413
DEBUG    ┝━ 🐛 [debug]: Computed 0 FRI layers from composition polynomial evaluations in 0 ms | log.target: "winter_prover" | log.module_path: "winter_prover" | log.file: "/Users/fumuran/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winter-prover-0.7.1/src/lib.rs" | log.line: 425
DEBUG    ┝━ 🐛 [debug]: Determined 26 unique query positions in 1 ms | log.target: "winter_prover" | log.module_path: "winter_prover" | log.file: "/Users/fumuran/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winter-prover-0.7.1/src/lib.rs" | log.line: 441
DEBUG    ┕━ 🐛 [debug]: Built proof object in 0 ms | log.target: "winter_prover" | log.module_path: "winter_prover" | log.file: "/Users/fumuran/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winter-prover-0.7.1/src/lib.rs" | log.line: 471
Program with hash 74184a73e531e907b5568906c1c2b53732bebcdcd7e197892694aba1d6ce197d proved in 15 ms
INFO     Writing data to proof file [ 226µs | 100.00% ] path: miden/examples/logger/logger.proof | size: "29 KB"
INFO     Writing data to output file [ 417µs | 100.00% ] path: miden/examples/logger/logger.outputs
Output: [3, 7, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
```

So here we can see that some of the `debug` lines got shorter (actually now I see that only first line got changed), but most of them stays unchanged, since the way they should be logged is specified in the Winterfell (we can see that their target is `winter_prover`).

bobbinth reviewed (approved):
Looks great! Thank you! I left some comments inline - but they are pretty small.

> There is no way we can disable printing the logging level at the beginning of the line (`INFO`, `WARN` etc.) without rewriting the whole tree-like output: these levels are placed in the core of this tree. Potentially I can rewrite the whole tree, but I'm not sure that it is worth it.

Yeah - what we have is fine, no need to spend more time on this.

> I shortened the unnecessarily detailed lines in the `DEBUG` output of the prover (the ones that were in the VM repo), but most of them are printed from the winterfell implementation of the prover. So to make them all short I need to implement Tracing in the winterfell too, but I'm not sure we want to do that now. Anyway, I can create an issue for that.

I'd like to see how the output of `prove` looks like but we probably don't need to do anything else here, and once Winterfell implements tracing, it will all look good again.

bobbinth commented on verifier/Cargo.toml:
```diff
@@ -22,5 +22,6 @@ std = ["air/std", "vm-core/std", "winter-verifier/std"]
 
 [dependencies]
 air = { package = "miden-air", path = "../air", version = "0.8", default-features = false }
+tracing = "0.1.40"
```
Should we change this to `tracing = { version = "0.1", default-features = false, optional = true }` as we do in other files?

bobbinth commented on docs/src/intro/usage.md:
````diff
@@ -83,6 +83,12 @@ For example:
 ./target/optimized/miden prove --help
 ```
 
+It is possible to specify the logging level at the beginning of the script. To set the level the value of the `MIDEN_LOG` environment variable should be specified. For example:
````
I would put a a sub-header above this and change the text as follows:
```
#### Enabling logging
You can use `MIDEN_LOG` environment variable to control how much logging output the VM produces. For example:
```

bobbinth reviewed (approved):
Looks great! Thank you! Just a couple of minor comments and we can merge.
